packages/@mantine/dates/src/components/TimeInput/TimeInput.tsx

```diff
@@ -33,14 +39,69 @@ const defaultProps: Partial<TimeInputProps> = {};
 
 export const TimeInput = factory<TimeInputFactory>((_props, ref) => {
   const props = useProps('TimeInput', defaultProps, _props);
-  const { classNames, styles, unstyled, vars, withSeconds, ...others } = props;
+  const {
+    classNames,
+    styles,
+    unstyled,
+    vars,
+    withSeconds,
+    minTime,
+    maxTime,
+    value,
+    onChange,
+    ...others
+  } = props;
 
   const { resolvedClassNames, resolvedStyles } = useResolvedStylesApi<TimeInputFactory>({
     classNames,
     styles,
     props,
   });
 
+  const onTimeChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    if (minTime !== undefined && maxTime !== undefined) {
+      const val = event.currentTarget.value;
+
+      if (val) {
+        const [hours, minutes, seconds] = val.split(':').map(Number);
+        const isValid =
+          (hours.toString().length === 2 || hours === 0) &&
+          (minutes.toString().length === 2 || minutes === 0) &&
+          (withSeconds ? seconds.toString().length === 2 || seconds === 0 : true);
+
+        if (isValid) {
+          if (minTime) {
+            const [minHours, minMinutes, minSeconds] = minTime.split(':').map(Number);
+
+            if (
+              hours < minHours ||
+              (hours === minHours && minutes < minMinutes) ||
+              (withSeconds && hours === minHours && minutes === minMinutes && seconds < minSeconds)
+            ) {
+              event.currentTarget.value = minTime;
+            }
+          }
+
+          if (maxTime) {
+            const [maxHours, maxMinutes, maxSeconds] = maxTime.split(':').map(Number);
+
+            if (
+              hours > maxHours ||
+              (hours === maxHours && minutes > maxMinutes) ||
+              (withSeconds && hours === maxHours && minutes === maxMinutes && seconds > maxSeconds)
+            ) {
+              event.currentTarget.value = maxTime;
+            }
+          }
+        }
+      }
+    }
```

Comment:
I'm not 100% sure this is easily readable, the method can be split into small functions if requested.